cli/tsc_config.rs

```diff
@@ -0,0 +1,238 @@
+  "declarationMap",
+  "diagnostics",
+  "downlevelIteration",
+  "emitBOM",
+  "emitDeclarationOnly",
+  "esModuleInterop",
+  "extendedDiagnostics",
+  "forceConsistentCasingInFileNames",
+  "generateCpuProfile",
+  "help",
+  "importHelpers",
+  "incremental",
+  "inlineSourceMap",
+  "inlineSources",
+  "init",
+  "listEmittedFiles",
+  "listFiles",
+  "mapRoot",
+  "maxNodeModuleJsDepth",
+  "module",
+  "moduleResolution",
+  "newLine",
+  "noEmit",
+  "noEmitHelpers",
+  "noEmitOnError",
+  "noLib",
+  "noResolve",
+  "out",
+  "outDir",
+  "outFile",
+  "paths",
+  "preserveConstEnums",
+  "preserveSymlinks",
+  "preserveWatchOutput",
+  "pretty",
+  "reactNamespace",
+  "resolveJsonModule",
+  "rootDir",
+  "rootDirs",
+  "showConfig",
+  "skipDefaultLibCheck",
+  "skipLibCheck",
+  "sourceMap",
+  "sourceRoot",
+  "stripInternal",
+  "target",
+  "traceResolution",
+  "tsBuildInfoFile",
+  "types",
+  "typeRoots",
+  "useDefineForClassFields",
+  "version",
+  "watch",
+];
+
+/// A function that works like JavaScript's `Object.assign()`.
+pub fn json_merge(a: &mut Value, b: &Value) {
+  match (a, b) {
+    (&mut Value::Object(ref mut a), &Value::Object(ref b)) => {
+      for (k, v) in b {
+        json_merge(a.entry(k.clone()).or_insert(Value::Null), v);
+      }
+    }
+    (a, b) => {
+      *a = b.clone();
+    }
+  }
+}
+
+/// Convert a jsonc libraries `JsonValue` to a serde `Value`.
+fn jsonc_to_serde(j: JsonValue) -> Value {
+  match j {
+    JsonValue::Array(arr) => {
+      let vec = arr.into_iter().map(jsonc_to_serde).collect();
+      Value::Array(vec)
+    }
+    JsonValue::Boolean(bool) => Value::Bool(bool),
+    JsonValue::Null => Value::Null,
+    JsonValue::Number(num) => {
+      let number =
+        serde_json::Number::from_str(&num).expect("could not parse number");
+      Value::Number(number)
+    }
+    JsonValue::Object(obj) => {
+      let mut map = serde_json::map::Map::new();
+      for (key, json_value) in obj.into_iter() {
+        map.insert(key, jsonc_to_serde(json_value));
+      }
+      Value::Object(map)
+    }
+    JsonValue::String(str) => Value::String(str),
+  }
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct TSConfigJson {
+  compiler_options: Option<HashMap<String, Value>>,
+  exclude: Option<Vec<String>>,
+  extends: Option<String>,
+  files: Option<Vec<String>>,
+  include: Option<Vec<String>>,
+  references: Option<Value>,
+  type_acquisition: Option<Value>,
+}
+
+pub fn parse_raw_config<T: AsRef<str>>(
+  config_text: T,
+) -> Result<Value, ErrBox> {
+  assert!(!config_text.as_ref().is_empty());
+  let jsonc = jsonc_parser::parse_to_value(config_text.as_ref())?.unwrap();
+  Ok(jsonc_to_serde(jsonc))
+}
+
+/// Take a string of JSONC, parse it and return a serde `Value` of the text.
+/// The result also contains any options that were ignored.
+pub fn parse_config<T: AsRef<str>>(
+  config_text: T,
+) -> Result<(Value, Option<IgnoredCompilerOptions>), ErrBox> {
```

Comment:
This is a stylistic nitpick but I prefer `&str` arguments over `AsRef<str>`... I think in some places it makes sense to be more generic but here there is just a single caller.

Reply:
👍  sure, done